File: antd-kotlin/lib/src/main/kotlin/com/autonomi/sdk/AntdRestClient.kt
package com.autonomi.sdk

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.put
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import okhttp3.Response
import java.time.Duration
import java.util.Base64

class AntdRestClient(
    baseUrl: String = "http://localhost:8082",
    timeout: Duration = Duration.ofSeconds(300),
) : IAntdClient {

    companion object {
        /**
         * Create a client by auto-discovering the daemon port from the
         * `daemon.port` file.  Falls back to `http://localhost:8082` if not found.
         */
        fun autoDiscover(timeout: Duration = Duration.ofSeconds(300)): AntdRestClient {
            val url = DaemonDiscovery.discoverDaemonUrl().ifEmpty { "http://localhost:8082" }
            return AntdRestClient(url, timeout)
        }
    }

    private val baseUrl = baseUrl.trimEnd('/')
    private val http = OkHttpClient.Builder()
        .callTimeout(timeout)
        .readTimeout(timeout)
        .writeTimeout(timeout)
        .build()
    private val json = Json { ignoreUnknownKeys = true }
    private val jsonMediaType = "application/json; charset=utf-8".toMediaType()

    override fun close() {
        http.dispatcher.executorService.shutdown()
        http.connectionPool.evictAll()
    }

    // ── Helpers ──

    private suspend inline fun <reified T> getJson(path: String): T = withContext(Dispatchers.IO) {
        val request = Request.Builder().url("$baseUrl$path").get().build()
        val response = http.newCall(request).execute()
        ensureSuccess(response)
        json.decodeFromString<T>(( response.body?.string() ?: "" ))
    }

    private suspend inline fun <reified T> postJson(path: String, body: String): T = withContext(Dispatchers.IO) {
        val request = Request.Builder()
            .url("$baseUrl$path")
            .post(body.toRequestBody(jsonMediaType))
            .build()
        val response = http.newCall(request).execute()
        ensureSuccess(response)
        json.decodeFromString<T>(( response.body?.string() ?: "" ))
    }

    private suspend fun postJsonNoResult(path: String, body: String) = withContext(Dispatchers.IO) {
        val request = Request.Builder()
            .url("$baseUrl$path")
            .post(body.toRequestBody(jsonMediaType))
            .build()
        val response = http.newCall(request).execute()
        ensureSuccess(response)
    }

    private fun ensureSuccess(response: Response) {
        if (response.isSuccessful) return
        val body = response.body?.string() ?: ""
        throw ExceptionMapping.fromHttpStatus(response.code, body)
    }

    private fun b64(data: ByteArray): String = Base64.getEncoder().encodeToString(data)
    private fun fromB64(s: String): ByteArray = Base64.getDecoder().decode(s)

    // ── Health ──

    override suspend fun health(): HealthStatus = try {
        getJson<HealthResponseDto>("/health").toHealthStatus()
    } catch (_: Exception) {
        HealthStatus(false, "unknown")
    }

    // ── Data ──

    override suspend fun dataPutPublic(data: ByteArray, paymentMode: PaymentMode): DataPutPublicResult {
        val body = buildJsonObject {
            put("data", b64(data))
            put("payment_mode", paymentMode.wire)
        }.toString()
        val resp = postJson<DataPutPublicDto>("/v1/data/public", body)
        return DataPutPublicResult(resp.address, resp.chunksStored, resp.paymentModeUsed)
    }

    override suspend fun dataGetPublic(address: String): ByteArray {
        val resp = getJson<DataGetDto>("/v1/data/public/$address")
        return fromB64(resp.data)
    }

    override suspend fun dataPut(data: ByteArray, paymentMode: PaymentMode): DataPutResult {
        val body = buildJsonObject {
            put("data", b64(data))
            put("payment_mode", paymentMode.wire)
        }.toString()
        val resp = postJson<DataPutDto>("/v1/data", body)
        return DataPutResult(resp.dataMap, resp.chunksStored, resp.paymentModeUsed)
    }

    override suspend fun dataGet(dataMap: String): ByteArray {
        val body = buildJsonObject { put("data_map", dataMap) }.toString()
        val resp = postJson<DataGetDto>("/v1/data/get", body)
        return fromB64(resp.data)
    }

    override suspend fun dataCost(data: ByteArray, paymentMode: PaymentMode): UploadCostEstimate {
        val body = buildJsonObject {
            put("data", b64(data))
            put("payment_mode", paymentMode.wire)
        }.toString()
        val resp = postJson<CostDto>("/v1/data/cost", body)
        return UploadCostEstimate(resp.cost, resp.fileSize, resp.chunkCount, resp.estimatedGasCostWei, resp.paymentMode)
    }

    // ── Chunks ──

    override suspend fun chunkPut(data: ByteArray): PutResult {
        val body = buildJsonObject { put("data", b64(data)) }.toString()
        val resp = postJson<ChunkPutDto>("/v1/chunks", body)
        return PutResult(resp.cost, resp.address)
    }

    override suspend fun chunkGet(address: String): ByteArray {
        val resp = getJson<DataGetDto>("/v1/chunks/$address")
        return fromB64(resp.data)
    }

    /**
     * Prepares a single chunk for external-signer publish via
     * `POST /v1/chunks/prepare`.
     *
     * Returns either `alreadyStored=true` (no payment needed, finalize is
     * unnecessary) or a wave-batch payment intent. After the external signer
     * pays via `payForQuotes`, call [finalizeChunkUpload] with the resulting
     * tx hashes.
     *
     * Requires antd >= 0.7.0.
     */
    override suspend fun prepareChunkUpload(data: ByteArray): PrepareChunkResult {
        val body = buildJsonObject { put("data", b64(data)) }.toString()
        val resp = postJson<PrepareChunkDto>("/v1/chunks/prepare", body)
        val payments = resp.payments?.map {
            PaymentInfo(it.quoteHash, it.rewardsAddress, it.amount)
        } ?: emptyList()
        return PrepareChunkResult(
            address = resp.address,
            alreadyStored = resp.alreadyStored,
            uploadId = resp.uploadId ?: "",
            paymentType = resp.paymentType ?: "",
            payments = payments,
            totalAmount = resp.totalAmount ?: "",
            paymentVaultAddress = resp.paymentVaultAddress ?: "",
            paymentTokenAddress = resp.paymentTokenAddress ?: "",
            rpcUrl = resp.rpcUrl ?: "",
        )
    }

    /**
     * Submits a prepared single chunk to the network after the external signer
     * has paid, via `POST /v1/chunks/finalize`.
     *
     * Returns the network address of the stored chunk (matches
     * [PrepareChunkResult.address]).
     *
     * Requires antd >= 0.7.0.
     */
    override suspend fun finalizeChunkUpload(uploadId: String, txHashes: Map<String, String>): String {
        val body = buildJsonObject {
            put("upload_id", uploadId)
            put("tx_hashes", buildJsonObject {
                txHashes.forEach { (k, v) -> put(k, v) }
            })
        }.toString()
        val resp = postJson<FinalizeChunkDto>("/v1/chunks/finalize", body)
        return resp.address
    }

    // ── Files ──

    override suspend fun filePutPublic(path: String, paymentMode: PaymentMode): FilePutPublicResult {
        val body = buildJsonObject {
            put("path", path)
            put("payment_mode", paymentMode.wire)
        }.toString()
        val resp = postJson<FilePutPublicDto>("/v1/files/public", body)
        return FilePutPublicResult(
            address = resp.address,
            storageCostAtto = resp.storageCostAtto,
            gasCostWei = resp.gasCostWei,
            chunksStored = resp.chunksStored,
            paymentModeUsed = resp.paymentModeUsed,
        )
    }

    override suspend fun fileGetPublic(address: String, destPath: String) {
        val body = buildJsonObject {
            put("address", address)
            put("dest_path", destPath)
        }.toString()
        postJsonNoResult("/v1/files/public/get", body)
    }

    override suspend fun filePut(path: String, paymentMode: PaymentMode): FilePutResult {
        val body = buildJsonObject {
            put("path", path)
            put("payment_mode", paymentMode.wire)
        }.toString()
        val resp = postJson<FilePutDto>("/v1/files", body)
        return FilePutResult(
            dataMap = resp.dataMap,
            storageCostAtto = resp.storageCostAtto,
            gasCostWei = resp.gasCostWei,
            chunksStored = resp.chunksStored,
            paymentModeUsed = resp.paymentModeUsed,
        )
    }

    override suspend fun fileGet(dataMap: String, destPath: String) {
        val body = buildJsonObject {
            put("data_map", dataMap)
            put("dest_path", destPath)
        }.toString()
        postJsonNoResult("/v1/files/get", body)
    }

    override suspend fun fileCost(path: String, isPublic: Boolean, paymentMode: PaymentMode): UploadCostEstimate {
        val body = buildJsonObject {
            put("path", path)
            put("is_public", isPublic)
            put("payment_mode", paymentMode.wire)
        }.toString()
        val resp = postJson<CostDto>("/v1/files/cost", body)
        return UploadCostEstimate(resp.cost, resp.fileSize, resp.chunkCount, resp.estimatedGasCostWei, resp.paymentMode)
    }

    // ── Wallet ──

    override suspend fun walletAddress(): WalletAddress {
        val resp = getJson<WalletAddressDto>("/v1/wallet/address")
        return WalletAddress(resp.address)
    }

    override suspend fun walletBalance(): WalletBalance {
        val resp = getJson<WalletBalanceDto>("/v1/wallet/balance")
        return WalletBalance(resp.balance, resp.gasBalance)
    }

    /** Approves the wallet to spend tokens on payment contracts (one-time operation). */
    override suspend fun walletApprove(): Boolean {
        val body = buildJsonObject {}.toString()
        val resp = postJson<WalletApproveDto>("/v1/wallet/approve", body)
        return resp.approved
    }

    // ── External Signer (Two-Phase Upload) ──

    /**
     * Prepares a file upload for external signing.
     *
     * Pass `visibility = "public"` to bundle the DataMap chunk into the same
     * external-signer payment batch — the resulting `dataMapAddress` on
     * finalize is the shareable retrieval handle. `null` (default) or
     * `"private"` keeps the existing private-only behaviour.
     */
    override suspend fun prepareUpload(path: String, visibility: String?): PrepareUploadResult {
        val body = buildJsonObject {
            put("path", path)
            if (visibility != null) put("visibility", visibility)
        }.toString()
        val resp = postJson<PrepareUploadDto>("/v1/upload/prepare", body)
        return mapPrepareUpload(resp)
    }

    /**
     * Convenience wrapper: prepares a *public* file upload for external
     * signing. Equivalent to [prepareUpload] with `visibility="public"`.
     *
     * Requires antd >= 0.6.1.
     */
    override suspend fun prepareUploadPublic(path: String): PrepareUploadResult =
        prepareUpload(path, visibility = "public")

    /**
     * Prepares a data upload for external signing.
     *
     * Note: `visibility="public"` returns 501 from the daemon until upstream
     * `data_prepare_upload_with_visibility` lands; use [prepareUploadPublic]
     * with a file path until then.
     */
    override suspend fun prepareDataUpload(data: ByteArray, visibility: String?): PrepareUploadResult {
        val body = buildJsonObject {
            put("data", b64(data))
            if (visibility != null) put("visibility", visibility)
        }.toString()
        val resp = postJson<PrepareUploadDto>("/v1/data/prepare", body)
        return mapPrepareUpload(resp)
    }

    /** Finalizes an upload after an external signer has submitted payment transactions. */
    override suspend fun finalizeUpload(uploadId: String, txHashes: Map<String, String>): FinalizeUploadResult {
        val body = buildJsonObject {
            put("upload_id", uploadId)
            put("tx_hashes", buildJsonObject {
                txHashes.forEach { (k, v) -> put(k, v) }
            })
        }.toString()
        val resp = postJson<FinalizeUploadDto>("/v1/upload/finalize", body)
        return FinalizeUploadResult(
            address = resp.address,
            chunksStored = resp.chunksStored,
            dataMap = resp.dataMap,
            dataMapAddress = resp.dataMapAddress,
        )
    }

    /** Finalizes a merkle batch upload by selecting a winner pool. */
    override suspend fun finalizeMerkleUpload(uploadId: String, winnerPoolHash: String): FinalizeMerkleUploadResult {
        val body = buildJsonObject {
            put("upload_id", uploadId)
            put("winner_pool_hash", winnerPoolHash)
        }.toString()
        val resp = postJson<FinalizeUploadDto>("/v1/upload/finalize", body)
        return FinalizeMerkleUploadResult(
            address = resp.address,
            chunksStored = resp.chunksStored,
            dataMap = resp.dataMap,
            dataMapAddress = resp.dataMapAddress,
        )
    }

    private fun mapPrepareUpload(resp: PrepareUploadDto): PrepareUploadResult {
        val payments = resp.payments?.map { PaymentInfo(it.quoteHash, it.rewardsAddress, it.amount) } ?: emptyList()
        val poolCommitments = resp.poolCommitments?.map { pc ->
            PoolCommitmentEntry(pc.poolHash, pc.candidates.map { CandidateNodeEntry(it.rewardsAddress, it.amount) })
        }
        return PrepareUploadResult(
            uploadId = resp.uploadId,
            payments = payments,
            totalAmount = resp.totalAmount,
            paymentVaultAddress = resp.paymentVaultAddress,
            paymentTokenAddress = resp.paymentTokenAddress,
            rpcUrl = resp.rpcUrl,
            paymentType = resp.paymentType ?: "wave_batch",
            depth = resp.depth,
            poolCommitments = poolCommitments,
            merklePaymentTimestamp = resp.merklePaymentTimestamp,
        )
    }
}

File: antd-swift/Sources/AntdSdk/AntdRestClient.swift
import Foundation
#if canImport(FoundationNetworking)
import FoundationNetworking
#endif

public final class AntdRestClient: AntdClientProtocol, @unchecked Sendable {

    private let baseURL: String
    private let session: URLSession

    public init(baseURL: String = "http://localhost:8082", timeout: TimeInterval = 300) {
        self.baseURL = baseURL.trimmingCharacters(in: CharacterSet(charactersIn: "/"))
        let config = URLSessionConfiguration.default
        config.timeoutIntervalForRequest = timeout
        config.timeoutIntervalForResource = timeout
        self.session = URLSession(configuration: config)
    }

    /// Internal init for tests that want to inject a custom session
    /// (typically with a stub `URLProtocol` registered via
    /// `configuration.protocolClasses`).
    internal init(baseURL: String, session: URLSession) {
        self.baseURL = baseURL.trimmingCharacters(in: CharacterSet(charactersIn: "/"))
        self.session = session
    }

    // MARK: - Helpers

    private func getJSON<T: Decodable>(_ path: String) async throws -> T {
        guard let url = URL(string: "\(baseURL)\(path)") else { throw BadRequestError("invalid URL: \(baseURL)\(path)") }
        let (data, response) = try await session.data(from: url)
        try ensureSuccess(response, data: data)
        return try JSONDecoder.snakeCase.decode(T.self, from: data)
    }

    private func postJSON<T: Decodable>(_ path: String, body: Any) async throws -> T {
        guard let url = URL(string: "\(baseURL)\(path)") else { throw BadRequestError("invalid URL: \(baseURL)\(path)") }
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.httpBody = try JSONSerialization.data(withJSONObject: body)
        let (data, response) = try await session.data(for: request)
        try ensureSuccess(response, data: data)
        return try JSONDecoder.snakeCase.decode(T.self, from: data)
    }

    private func postJSONNoResult(_ path: String, body: Any) async throws {
        guard let url = URL(string: "\(baseURL)\(path)") else { throw BadRequestError("invalid URL: \(baseURL)\(path)") }
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.httpBody = try JSONSerialization.data(withJSONObject: body)
        let (data, response) = try await session.data(for: request)
        try ensureSuccess(response, data: data)
    }

    private func headExists(_ path: String) async throws -> Bool {
        guard let url = URL(string: "\(baseURL)\(path)") else { throw BadRequestError("invalid URL: \(baseURL)\(path)") }
        var request = URLRequest(url: url)
        request.httpMethod = "HEAD"
        let (data, response) = try await session.data(for: request)
        guard let httpResponse = response as? HTTPURLResponse else { return false }
        if httpResponse.statusCode == 404 { return false }
        try ensureSuccess(response, data: data)
        return true
    }

    private func ensureSuccess(_ response: URLResponse, data: Data) throws {
        guard let httpResponse = response as? HTTPURLResponse else { return }
        guard (200...299).contains(httpResponse.statusCode) else {
            let body = String(data: data, encoding: .utf8) ?? ""
            throw ErrorMapping.fromHTTPStatus(httpResponse.statusCode, body: body)
        }
    }

    // MARK: - Health

    public func health() async throws -> HealthStatus {
        do {
            let resp: HealthResponseDTO = try await getJSON("/health")
            return resp.toHealthStatus()
        } catch {
            return HealthStatus(ok: false, network: "unknown")
        }
    }

    // MARK: - Data

    public func dataPut(_ data: Data, paymentMode: PaymentMode = .auto) async throws -> DataPutResult {
        let body: [String: Any] = [
            "data": data.base64EncodedString(),
            "payment_mode": paymentMode.rawValue,
        ]
        let resp: DataPutDTO = try await postJSON("/v1/data", body: body)
        return DataPutResult(
            dataMap: resp.dataMap,
            chunksStored: resp.chunksStored ?? 0,
            paymentModeUsed: resp.paymentModeUsed ?? ""
        )
    }

    public func dataGet(dataMap: String) async throws -> Data {
        let resp: DataDTO = try await postJSON("/v1/data/get", body: ["data_map": dataMap])
        guard let decoded = Data(base64Encoded: resp.data) else { throw BadRequestError("Invalid base64 data") }
        return decoded
    }

    public func dataPutPublic(_ data: Data, paymentMode: PaymentMode = .auto) async throws -> DataPutPublicResult {
        let body: [String: Any] = [
            "data": data.base64EncodedString(),
            "payment_mode": paymentMode.rawValue,
        ]
        let resp: DataPutPublicDTO = try await postJSON("/v1/data/public", body: body)
        return DataPutPublicResult(
            address: resp.address,
            chunksStored: resp.chunksStored ?? 0,
            paymentModeUsed: resp.paymentModeUsed ?? ""
        )
    }

    public func dataGetPublic(address: String) async throws -> Data {
        let resp: DataDTO = try await getJSON("/v1/data/public/\(address)")
        guard let decoded = Data(base64Encoded: resp.data) else { throw BadRequestError("Invalid base64 data") }
        return decoded
    }

    public func dataCost(_ data: Data, paymentMode: PaymentMode = .auto) async throws -> UploadCostEstimate {
        let body: [String: Any] = [
            "data": data.base64EncodedString(),
            "payment_mode": paymentMode.rawValue,
        ]
        let resp: CostDTO = try await postJSON("/v1/data/cost", body: body)
        return UploadCostEstimate(
            cost: resp.cost,
            fileSize: resp.fileSize ?? 0,
            chunkCount: resp.chunkCount ?? 0,
            estimatedGasCostWei: resp.estimatedGasCostWei ?? "",
            paymentMode: resp.paymentMode ?? "")
    }

    // MARK: - Chunks

    public func chunkPut(_ data: Data) async throws -> PutResult {
        let resp: CostAddressDTO = try await postJSON("/v1/chunks", body: ["data": data.base64EncodedString()])
        return PutResult(cost: resp.cost ?? "", address: resp.address)
    }

    public func chunkGet(address: String) async throws -> Data {
        let resp: DataDTO = try await getJSON("/v1/chunks/\(address)")
        guard let decoded = Data(base64Encoded: resp.data) else { throw BadRequestError("Invalid base64 data") }
        return decoded
    }

    /// Prepare a single chunk for external-signer publish.
    ///
    /// Returns either ``PrepareChunkResult/alreadyStored`` `= true` (no
    /// payment needed) or a wave-batch payment intent. After the external
    /// signer pays, call ``finalizeChunkUpload(uploadId:txHashes:)`` with the
    /// resulting tx hashes.
    public func prepareChunkUpload(_ data: Data) async throws -> PrepareChunkResult {
        let resp: PrepareChunkDTO = try await postJSON(
            "/v1/chunks/prepare",
            body: ["data": data.base64EncodedString()]
        )
        let payments = (resp.payments ?? []).map {
            PaymentInfo(quoteHash: $0.quoteHash, rewardsAddress: $0.rewardsAddress, amount: $0.amount)
        }
        return PrepareChunkResult(
            address: resp.address,
            alreadyStored: resp.alreadyStored ?? false,
            uploadId: resp.uploadId ?? "",
            paymentType: resp.paymentType ?? "",
            payments: payments,
            totalAmount: resp.totalAmount ?? "",
            paymentVaultAddress: resp.paymentVaultAddress ?? "",
            paymentTokenAddress: resp.paymentTokenAddress ?? "",
            rpcUrl: resp.rpcUrl ?? ""
        )
    }

    /// Submit a prepared chunk to the network after external payment.
    /// Returns the network address of the stored chunk (matches
    /// ``PrepareChunkResult/address``).
    public func finalizeChunkUpload(uploadId: String, txHashes: [String: String]) async throws -> String {
        let body: [String: Any] = ["upload_id": uploadId, "tx_hashes": txHashes]
        let resp: AddressDTO = try await postJSON("/v1/chunks/finalize", body: body)
        return resp.address
    }

    // MARK: - Files

    public func filePut(path: String, paymentMode: PaymentMode = .auto) async throws -> FilePutResult {
        let body: [String: Any] = [
            "path": path,
            "payment_mode": paymentMode.rawValue,
        ]
        let resp: FilePutDTO = try await postJSON("/v1/files", body: body)
        return FilePutResult(
            dataMap: resp.dataMap,
            storageCostAtto: resp.storageCostAtto,
            gasCostWei: resp.gasCostWei,
            chunksStored: resp.chunksStored,
            paymentModeUsed: resp.paymentModeUsed
        )
    }

    public func fileGet(dataMap: String, destPath: String) async throws {
        try await postJSONNoResult("/v1/files/get", body: ["data_map": dataMap, "dest_path": destPath])
    }

    public func filePutPublic(path: String, paymentMode: PaymentMode = .auto) async throws -> FilePutPublicResult {
        let body: [String: Any] = [
            "path": path,
            "payment_mode": paymentMode.rawValue,
        ]
        let resp: FilePutPublicDTO = try await postJSON("/v1/files/public", body: body)
        return FilePutPublicResult(
            address: resp.address,
            storageCostAtto: resp.storageCostAtto,
            gasCostWei: resp.gasCostWei,
            chunksStored: resp.chunksStored,
            paymentModeUsed: resp.paymentModeUsed
        )
    }

    public func fileGetPublic(address: String, destPath: String) async throws {
        try await postJSONNoResult("/v1/files/public/get", body: ["address": address, "dest_path": destPath])
    }

    public func fileCost(path: String, isPublic: Bool = true, paymentMode: PaymentMode = .auto) async throws -> UploadCostEstimate {
        let body: [String: Any] = [
            "path": path,
            "is_public": isPublic,
            "payment_mode": paymentMode.rawValue,
        ]
        let resp: CostDTO = try await postJSON("/v1/files/cost", body: body)
        return UploadCostEstimate(
            cost: resp.cost,
            fileSize: resp.fileSize ?? 0,
            chunkCount: resp.chunkCount ?? 0,
            estimatedGasCostWei: resp.estimatedGasCostWei ?? "",
            paymentMode: resp.paymentMode ?? "")
    }

    // MARK: - Wallet

    public func walletAddress() async throws -> WalletAddress {
        let resp: WalletAddressDTO = try await getJSON("/v1/wallet/address")
        return WalletAddress(address: resp.address)
    }

    public func walletBalance() async throws -> WalletBalance {
        let resp: WalletBalanceDTO = try await getJSON("/v1/wallet/balance")
        return WalletBalance(balance: resp.balance, gasBalance: resp.gasBalance)
    }

    /// Approves the wallet to spend tokens on payment contracts (one-time operation).
    public func walletApprove() async throws -> Bool {
        let resp: WalletApproveDTO = try await postJSON("/v1/wallet/approve", body: [:] as [String: Any])
        return resp.approved
    }

    // MARK: - External Signer (Two-Phase Upload)

    /// Prepares a file upload for external signing.
    ///
    /// - Parameters:
    ///   - path: Path to the file to upload.
    ///   - visibility: ``"public"`` bundles the DataMap chunk into the same
    ///     external-signer payment batch — the resulting
    ///     ``FinalizeUploadResult/dataMapAddress`` on finalize is the
    ///     shareable retrieval handle. ``"private"`` or `nil` keeps the
    ///     existing private-only behaviour. The field is omitted from the
    ///     wire request when `nil`, preserving compatibility with daemons
    ///     that predate the public-prepare wire shape.
    public func prepareUpload(path: String, visibility: String? = nil) async throws -> PrepareUploadResult {
        var body: [String: Any] = ["path": path]
        if let visibility = visibility { body["visibility"] = visibility }
        let resp: PrepareUploadDTO = try await postJSON("/v1/upload/prepare", body: body)
        return mapPrepareDTO(resp)
    }

    /// Convenience wrapper: prepare a *public* file upload for external signing.
    /// Equivalent to ``prepareUpload(path:visibility:)`` with `visibility: "public"`.
    public func prepareUploadPublic(path: String) async throws -> PrepareUploadResult {
        try await prepareUpload(path: path, visibility: "public")
    }

    /// Prepares a data upload for external signing.
    /// Takes raw bytes, base64-encodes them, and POSTs to /v1/data/prepare.
    public func prepareDataUpload(_ data: Data) async throws -> PrepareUploadResult {
        let resp: PrepareUploadDTO = try await postJSON("/v1/data/prepare", body: ["data": data.base64EncodedString()])
        return mapPrepareDTO(resp)
    }

    /// Finalizes a wave-batch upload after an external signer has submitted payment transactions.
    public func finalizeUpload(uploadId: String, txHashes: [String: String]) async throws -> FinalizeUploadResult {
        let body: [String: Any] = ["upload_id": uploadId, "tx_hashes": txHashes]
        let resp: FinalizeUploadDTO = try await postJSON("/v1/upload/finalize", body: body)
        return FinalizeUploadResult(
            address: resp.address ?? "",
            chunksStored: resp.chunksStored,
            dataMap: resp.dataMap ?? "",
            dataMapAddress: resp.dataMapAddress ?? ""
        )
    }

    /// Finalizes a merkle batch upload after the external signer has submitted
    /// the `payForMerkleTree` transaction. `winnerPoolHash` is the bytes32 value
    /// from the `MerklePaymentMade` event (hex with 0x prefix).
    public func finalizeMerkleUpload(uploadId: String, winnerPoolHash: String) async throws -> FinalizeMerkleUploadResult {
        let body: [String: Any] = ["upload_id": uploadId, "winner_pool_hash": winnerPoolHash]
        let resp: FinalizeUploadDTO = try await postJSON("/v1/upload/finalize", body: body)
        return FinalizeMerkleUploadResult(
            address: resp.address ?? "",
            chunksStored: resp.chunksStored,
            dataMap: resp.dataMap ?? "",
            dataMapAddress: resp.dataMapAddress ?? ""
        )
    }

    // MARK: - Prepare DTO Mapping

    private func mapPrepareDTO(_ resp: PrepareUploadDTO) -> PrepareUploadResult {
        let payments = (resp.payments ?? []).map { PaymentInfo(quoteHash: $0.quoteHash, rewardsAddress: $0.rewardsAddress, amount: $0.amount) }
        let poolCommitments = resp.poolCommitments?.map { pc in
            PoolCommitmentEntry(
                poolHash: pc.poolHash,
                candidates: pc.candidates.map { CandidateNodeEntry(rewardsAddress: $0.rewardsAddress, amount: $0.amount) }
            )
        }
        return PrepareUploadResult(
            uploadId: resp.uploadId,
            payments: payments,
            totalAmount: resp.totalAmount,
            paymentVaultAddress: resp.paymentVaultAddress,
            paymentTokenAddress: resp.paymentTokenAddress,
            rpcUrl: resp.rpcUrl,
            paymentType: resp.paymentType ?? "wave_batch",
            depth: resp.depth,
            poolCommitments: poolCommitments,
            merklePaymentTimestamp: resp.merklePaymentTimestamp,
            totalChunks: resp.totalChunks ?? 0,
            alreadyStoredCount: resp.alreadyStoredCount ?? 0
        )
    }
}

// MARK: - Internal DTOs

private struct HealthResponseDTO: Decodable {
    let status: String?
    let network: String?
    let version: String?
    let evm_network: String?
    let uptime_seconds: UInt64?
    let build_commit: String?
    let payment_token_address: String?
    let payment_vault_address: String?

    func toHealthStatus() -> HealthStatus {
        HealthStatus(
            ok: status == "ok",
            network: network ?? "unknown",
            version: version ?? "",
            evmNetwork: evm_network ?? "",
            uptimeSeconds: uptime_seconds ?? 0,
            buildCommit: build_commit ?? "",
            paymentTokenAddress: payment_token_address ?? "",
            paymentVaultAddress: payment_vault_address ?? ""
        )
    }
}

private struct CostAddressDTO: Decodable {
    // PUT responses sometimes omit `cost`; default to empty downstream (#69 in PR queue).
    let cost: String?
    let address: String
}

// Wire fields are snake_case; the shared JSONDecoder uses .convertFromSnakeCase,
// so property names are camelCase here. Adding explicit CodingKeys would
// suppress that strategy (the decoder skips the conversion when CodingKeys are
// present and matches against the raw values verbatim) and silently nil out
// every snake_case-named field.
private struct DataPutDTO: Decodable {
    let dataMap: String
    let chunksStored: UInt64?
    let paymentModeUsed: String?
}

private struct DataPutPublicDTO: Decodable {
    let address: String
    let chunksStored: UInt64?
    let paymentModeUsed: String?
}

private struct FilePutDTO: Decodable {
    let dataMap: String
    let storageCostAtto: String
    let gasCostWei: String
    let chunksStored: UInt64
    let paymentModeUsed: String
}

private struct FilePutPublicDTO: Decodable {
    let address: String
    let storageCostAtto: String
    let gasCostWei: String
    let chunksStored: UInt64
    let paymentModeUsed: String
}

private struct DataDTO: Decodable {
    let data: String
}

private struct CostDTO: Decodable {
    let cost: String
    // Wire fields are snake_case (file_size, chunk_count, …); the shared
    // JSONDecoder uses .convertFromSnakeCase, so name them camelCase here
    // — otherwise decoding silently nils them out and the caller sees zeros.
    let fileSize: UInt64?
    let chunkCount: UInt32?
    let estimatedGasCostWei: String?
    let paymentMode: String?
}

private struct WalletAddressDTO: Decodable {
    let address: String
}

private struct WalletBalanceDTO: Decodable {
    let balance: String
    let gasBalance: String
}

private struct WalletApproveDTO: Decodable {
    let approved: Bool
}

private struct PaymentInfoDTO: Decodable {
    let quoteHash: String
    let rewardsAddress: String
    let amount: String
}

private struct CandidateNodeEntryDTO: Decodable {
    let rewardsAddress: String
    let amount: String
}

private struct PoolCommitmentEntryDTO: Decodable {
    let poolHash: String
    let candidates: [CandidateNodeEntryDTO]
}

private struct PrepareUploadDTO: Decodable {
    let uploadId: String
    let payments: [PaymentInfoDTO]?
    let totalAmount: String
    let paymentVaultAddress: String
    let paymentTokenAddress: String
    let rpcUrl: String
    let paymentType: String?
    let depth: Int?
    let poolCommitments: [PoolCommitmentEntryDTO]?
    let merklePaymentTimestamp: UInt64?
    // Optional so older daemons that omit these still decode (defaulted to 0 on map).
    let totalChunks: UInt64?
    let alreadyStoredCount: UInt64?
}

private struct FinalizeUploadDTO: Decodable {
    let address: String?
    let chunksStored: Int64
    // `data_map` is always present on success but old daemons may omit it;
    // `data_map_address` is populated only when prepare was public.
    let dataMap: String?
    let dataMapAddress: String?
}

private struct PrepareChunkDTO: Decodable {
    let address: String
    let alreadyStored: Bool?
    let uploadId: String?
    let paymentType: String?
    let payments: [PaymentInfoDTO]?
    let totalAmount: String?
    let paymentVaultAddress: String?
    let paymentTokenAddress: String?
    let rpcUrl: String?
}

private struct AddressDTO: Decodable {
    let address: String
}

extension JSONDecoder {
    static let snakeCase: JSONDecoder = {
        let decoder = JSONDecoder()
        decoder.keyDecodingStrategy = .convertFromSnakeCase
        return decoder
    }()
}
